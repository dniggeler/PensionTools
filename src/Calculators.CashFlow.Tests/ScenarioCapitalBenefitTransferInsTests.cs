using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using PensionCoach.Tools.CommonTypes;
using PensionCoach.Tools.CommonTypes.Tax;
using PensionCoach.Tools.TaxComparison;
using Snapshooter.Xunit;
using Xunit;

namespace Calculators.CashFlow.Tests
{
    [Trait("Higher Level Calculators", "Transfer-Ins")]
    public class ScenarioCapitalBenefitTransferInsTests : IClassFixture<CashFlowFixture<ITaxScenarioCalculator>>
    {
        private readonly CashFlowFixture<ITaxScenarioCalculator> fixture;

        public ScenarioCapitalBenefitTransferInsTests(CashFlowFixture<ITaxScenarioCalculator> fixture)
        {
            this.fixture = fixture;
        }

        [Fact(DisplayName = "Single Transfer-In With Withdrawal in 10 Years")]
        public async Task Calculate_Single_TransferIn_With_Withdrawal_In_10_Years()
        {
            // given
            int calculationYear = 2022;
            int bfsMunicipalityId = 261;
            TaxPerson person = new()
            {
                Name = "Unit Test Purchase",
                CivilStatus = CivilStatus.Married,
                ReligiousGroupType = ReligiousGroupType.Other,
                PartnerReligiousGroupType = ReligiousGroupType.Other,
                TaxableIncome = 150_000,
                TaxableFederalIncome = 100_000,
                TaxableWealth = 500_000
            };

            TransferInCapitalBenefitsScenarioModel scenarioModel = new()
            {
                NetReturnRate = 0.0M,
                TransferIns = new List<SingleTransferInModel>
                {
                    new(10000, new DateTime(2022, 1, 1)),
                },
                WithCapitalBenefitWithdrawal = true,
                CapitalBenefitsBeforeWithdrawal = 800_000,
                Withdrawals = new List<SingleTransferInModel>
                {
                    new(0.5M, new DateTime(2032, 12, 31)),
                    new(0.5M, new DateTime(2033, 12, 31)),
                },
            };

            // when
            var result = await fixture.Calculator.TransferInCapitalBenefitsAsync(
                calculationYear, bfsMunicipalityId, person, scenarioModel);

            // then
            Snapshot.Match(result, opt => opt.IgnoreFields("$..Id"));
        }
    }
}
