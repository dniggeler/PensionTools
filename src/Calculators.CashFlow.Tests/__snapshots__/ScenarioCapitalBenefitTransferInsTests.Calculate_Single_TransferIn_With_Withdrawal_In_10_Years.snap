[
  {
    "State": "IsRight",
    "Right": {
      "StartingYear": 2022,
      "NumberOfPeriods": 12,
      "DeltaSeries": [
        {
          "Year": 2022,
          "Amount": 3199.900,
          "AccountType": "Exogenous"
        },
        {
          "Year": 2023,
          "Amount": 3215.23000,
          "AccountType": "Exogenous"
        },
        {
          "Year": 2024,
          "Amount": 3230.5600000,
          "AccountType": "Exogenous"
        },
        {
          "Year": 2025,
          "Amount": 3245.890000000,
          "AccountType": "Exogenous"
        },
        {
          "Year": 2026,
          "Amount": 3261.22000000000,
          "AccountType": "Exogenous"
        },
        {
          "Year": 2027,
          "Amount": 3268.88500000000000,
          "AccountType": "Exogenous"
        },
        {
          "Year": 2028,
          "Amount": 3275.45500000000000000,
          "AccountType": "Exogenous"
        },
        {
          "Year": 2029,
          "Amount": 3283.12000000000000000000,
          "AccountType": "Exogenous"
        },
        {
          "Year": 2030,
          "Amount": 3290.78500000000000000000000,
          "AccountType": "Exogenous"
        },
        {
          "Year": 2031,
          "Amount": 3297.35500000000000000000000,
          "AccountType": "Exogenous"
        },
        {
          "Year": 2032,
          "Amount": 2626.4250000000000000000000,
          "AccountType": "Exogenous"
        },
        {
          "Year": 2033,
          "Amount": 2433.3100000000000000000000,
          "AccountType": "Exogenous"
        }
      ],
      "BenchmarkSeries": [
        {
          "Year": 2022,
          "Amount": 473331.375,
          "AccountType": "Wealth"
        },
        {
          "Year": 2022,
          "Amount": 0.0,
          "AccountType": "OccupationalPension"
        },
        {
          "Year": 2023,
          "Amount": 446721.88000,
          "AccountType": "Wealth"
        },
        {
          "Year": 2023,
          "Amount": 0.0,
          "AccountType": "OccupationalPension"
        },
        {
          "Year": 2024,
          "Amount": 420171.5150000,
          "AccountType": "Wealth"
        },
        {
          "Year": 2024,
          "Amount": 0.0,
          "AccountType": "OccupationalPension"
        },
        {
          "Year": 2025,
          "Amount": 393678.090000000,
          "AccountType": "Wealth"
        },
        {
          "Year": 2025,
          "Amount": 0.0,
          "AccountType": "OccupationalPension"
        },
        {
          "Year": 2026,
          "Amount": 367243.79500000000,
          "AccountType": "Wealth"
        },
        {
          "Year": 2026,
          "Amount": 0.0,
          "AccountType": "OccupationalPension"
        },
        {
          "Year": 2027,
          "Amount": 340846.73000000000000,
          "AccountType": "Wealth"
        },
        {
          "Year": 2027,
          "Amount": 0.0,
          "AccountType": "OccupationalPension"
        },
        {
          "Year": 2028,
          "Amount": 314479.23000000000000000,
          "AccountType": "Wealth"
        },
        {
          "Year": 2028,
          "Amount": 0.0,
          "AccountType": "OccupationalPension"
        },
        {
          "Year": 2029,
          "Amount": 288140.20000000000000000000,
          "AccountType": "Wealth"
        },
        {
          "Year": 2029,
          "Amount": 0.0,
          "AccountType": "OccupationalPension"
        },
        {
          "Year": 2030,
          "Amount": 261829.64000000000000000000000,
          "AccountType": "Wealth"
        },
        {
          "Year": 2030,
          "Amount": 0.0,
          "AccountType": "OccupationalPension"
        },
        {
          "Year": 2031,
          "Amount": 235548.64500000000000000000000,
          "AccountType": "Wealth"
        },
        {
          "Year": 2031,
          "Amount": 0.0,
          "AccountType": "OccupationalPension"
        },
        {
          "Year": 2032,
          "Amount": 581516.05000000000000000000000,
          "AccountType": "Wealth"
        },
        {
          "Year": 2032,
          "Amount": 400000.0,
          "AccountType": "OccupationalPension"
        },
        {
          "Year": 2033,
          "Amount": 774905.24500000000000000000000,
          "AccountType": "Wealth"
        },
        {
          "Year": 2033,
          "Amount": 200000.00,
          "AccountType": "OccupationalPension"
        }
      ],
      "ScenarioSeries": [
        {
          "Year": 2022,
          "Amount": 466531.275,
          "AccountType": "Wealth"
        },
        {
          "Year": 2022,
          "Amount": 10000.0,
          "AccountType": "OccupationalPension"
        },
        {
          "Year": 2023,
          "Amount": 439937.11000,
          "AccountType": "Wealth"
        },
        {
          "Year": 2023,
          "Amount": 10000.0,
          "AccountType": "OccupationalPension"
        },
        {
          "Year": 2024,
          "Amount": 413402.0750000,
          "AccountType": "Wealth"
        },
        {
          "Year": 2024,
          "Amount": 10000.0,
          "AccountType": "OccupationalPension"
        },
        {
          "Year": 2025,
          "Amount": 386923.980000000,
          "AccountType": "Wealth"
        },
        {
          "Year": 2025,
          "Amount": 10000.0,
          "AccountType": "OccupationalPension"
        },
        {
          "Year": 2026,
          "Amount": 360505.01500000000,
          "AccountType": "Wealth"
        },
        {
          "Year": 2026,
          "Amount": 10000.0,
          "AccountType": "OccupationalPension"
        },
        {
          "Year": 2027,
          "Amount": 334115.61500000000000,
          "AccountType": "Wealth"
        },
        {
          "Year": 2027,
          "Amount": 10000.0,
          "AccountType": "OccupationalPension"
        },
        {
          "Year": 2028,
          "Amount": 307754.68500000000000000,
          "AccountType": "Wealth"
        },
        {
          "Year": 2028,
          "Amount": 10000.0,
          "AccountType": "OccupationalPension"
        },
        {
          "Year": 2029,
          "Amount": 281423.32000000000000000000,
          "AccountType": "Wealth"
        },
        {
          "Year": 2029,
          "Amount": 10000.0,
          "AccountType": "OccupationalPension"
        },
        {
          "Year": 2030,
          "Amount": 255120.42500000000000000000000,
          "AccountType": "Wealth"
        },
        {
          "Year": 2030,
          "Amount": 10000.0,
          "AccountType": "OccupationalPension"
        },
        {
          "Year": 2031,
          "Amount": 228846.00000000000000000000000,
          "AccountType": "Wealth"
        },
        {
          "Year": 2031,
          "Amount": 10000.0,
          "AccountType": "OccupationalPension"
        },
        {
          "Year": 2032,
          "Amount": 579142.47500000000000000000000,
          "AccountType": "Wealth"
        },
        {
          "Year": 2032,
          "Amount": 405000.0,
          "AccountType": "OccupationalPension"
        },
        {
          "Year": 2033,
          "Amount": 774838.55500000000000000000000,
          "AccountType": "Wealth"
        },
        {
          "Year": 2033,
          "Amount": 202500.00,
          "AccountType": "OccupationalPension"
        }
      ]
    },
    "Left": null
  }
]
